src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,224 @@
+    pub dst: Option<B::AccelerationStructure>,
+
+    // TODO: We could enforce the following the type system?
+    // - in both vulkan (via `VUID-VkAccelerationStructureBuildGeometryInfoKHR-type-03792`) and DX12 (via type system), all of the structs here must be the same variant.
+    // - blas must be triangles or aabbs, tlas must be instances
+    pub geometries: &'a [AccelerationStructureGeometry<'a, B>],
+    // Both APIs support "array" and "array of pointers", presumably to allow for cheap reuse of the geometry descriptors.
+    // pgeometries: &'a [&'a GeometryDesc<'a, B>],
+    pub scratch: Option<(&'a B::Buffer, Offset)>,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkGeometryFlagBitsKHR.html
+    pub struct AccelerationStructureGeometryFlags: u32 {
+        /// indicates that this geometry does not invoke the any-hit shaders even if present in a hit group.
+        const OPAQUE = 0x1;
+        /// indicates that the implementation must only call the any-hit shader a single time for each primitive in this geometry. If this bit is absent an implementation may invoke the any-hit shader more than once for this geometry.
+        const NO_DUPLICATE_ANY_HIT_INVOCATION = 0x2;
+    }
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureGeometry<'a, B: Backend> {
+    pub flags: AccelerationStructureGeometryFlags,
+    pub geometry: Geometry<'a, B>,
+}
+
+#[derive(Debug)]
+pub enum Geometry<'a, B: Backend> {
+    Triangles(GeometryTriangles<'a, B>),
+
+    /// TODO bikeshed capitalization of AABBs.
+    Aabbs(GeometryAabbs<'a, B>),
+
+    Instances(GeometryInstances<'a, B>),
+}
+
+// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryTrianglesDataKHR.html
+// for docs: Memory Safety (https://nvpro-samples.github.io/vk_raytracing_tutorial_KHR/#accelerationstructure)
+#[derive(Debug)]
+pub struct GeometryTriangles<'a, B: Backend> {
+    /// Both APIs require support for at least the following:
+    /// - `(R32_G32, Float)`
+    /// - `(R32_G32_B32, Float)`
+    /// - `(R16_G16, Float)`
+    /// - `(R16_G16_B16_A16, Float)`
+    /// - `(R16_G16, Inorm)`
+    /// - `(R16_G16_B16_A16, Inorm)`
+    /// VK could support more by querying `VK_FORMAT_FEATURE_ACCELERATION_STRUCTURE_VERTEX_BUFFER_BIT_KHR`, DX12 is not queryable? Note [the DX12 ray tracing spec](https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#d3d12_raytracing_geometry_triangles_desc) says it supports more than [the Win32 docs](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_geometry_triangles_desc).
+    pub vertex_format: Format,
+    pub vertex_buffer: &'a B::Buffer,
+    pub vertex_buffer_offset: Offset,
+    pub vertex_buffer_stride: Offset,
+
+    /// aka "vertex count"?
+    pub max_vertex: Offset,
+
+    // index format must be DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT
+    // can also be VK_INDEX_TYPE_NONE_KHR/DXGI_FORMAT_UNKNOWN
```

Comment:
Should it be `Option<IndexType>` then? Or is the whole `index_buffer: None` supposed to match `VK_INDEX_TYPE_NONE_KHR`?

Reply:
I believe `VK_INDEX_TYPE_NONE_KHR` is meant to signal that there's no index data (and it's just a triangles list), since `VkBuffer` doesn't have a null value to achieve the same result.

The valid usage requirements for `vkCmdBuildAccelerationStructuresKHR` hints at this but doesn't say it explicitly.
- https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03768
- https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03806
- https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03807
- https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03712